core/src/alpenglow_consensus/vote_certificate.rs

```diff
@@ -70,70 +72,78 @@ impl LegacyVoteCertificate {
 
-    fn new(
-        certificate_id: CertificateId,
-        transactions: Vec<Arc<VoteMessage>>,
-    ) -> Result<Self, CertificateError> {
-        let (aggregate_signature, bitmap) = aggregate_vote_signatures(transactions)?;
-        Ok(CertificateMessage {
+    fn new(certificate_id: CertificateId) -> Self {
+        CertificateMessage {
             certificate: certificate_id.into(),
-            signature: aggregate_signature,
-            bitmap,
-        })
+            signature: Signature::default(),
+            bitmap: BitVec::<u8, Lsb0>::repeat(false, VALIDATOR_BITMAP_U8_SIZE),
+        }
     }
 
     fn vote_count(&self) -> usize {
         self.bitmap.count_ones()
     }
-}
-
-fn aggregate_vote_signatures(
-    transactions: Vec<Arc<VoteMessage>>,
-) -> Result<(Signature, BitVec<u8, Lsb0>), CertificateError> {
-    if transactions.len() > MAXIMUM_VALIDATORS {
-        return Err(CertificateError::IndexOutOfBound);
-    }
 
-    let mut aggregate_signature = SignatureProjective::default();
-    let mut bitmap = BitVec::<u8, Lsb0>::repeat(false, VALIDATOR_BITMAP_U8_SIZE);
-
-    // TODO: signature aggregation can be done out-of-order;
-    // consider aggregating signatures separately in parallel
-    for transaction in transactions {
-        // aggregate the signature
-        let signature: SignatureProjective = transaction
-            .signature
-            .try_into()
-            .map_err(|_| CertificateError::InvalidSignature)?;
-        aggregate_signature.aggregate_with([&signature]);
-
-        // set bit-vector for the validator
-        //
-        // TODO: This only accounts for one type of vote. Update this after
-        // we have a base3 encoding implementation.
-        if bitmap.len() < transaction.rank as usize {
-            return Err(CertificateError::IndexOutOfBound);
+    fn aggregate<'a, T>(&mut self, messages: T) -> Result<(), CertificateError>
+    where
+        T: Iterator<Item = &'a Self::VoteTransaction>,
+        Self: 'a,
+    {
+        // TODO: signature aggregation can be done out-of-order;
+        // consider aggregating signatures separately in parallel
+        let mut current_signature_uncompressed = if self.signature == Signature::default() {
+            SignatureProjective::default()
+        } else {
+            SignatureProjective::try_from(self.signature)
+                .map_err(|_| CertificateError::InvalidSignature)?
+        };
+
+        // aggregate the votes
+        for vote_message in messages {
+            // set bit-vector for the validator
+            //
+            // TODO: This only accounts for one type of vote. Update this after
+            // we have a base3 encoding implementation.
+            if self.bitmap.len() < vote_message.rank as usize {
+                return Err(CertificateError::IndexOutOfBound);
+            }
+            if self.bitmap.get(vote_message.rank as usize).as_deref() == Some(&true) {
+                return Ok(());
```

Comment:
Wdyt about omitting this check or panicking? It seems `VotePool::add_vote` already checks for duplicate votes, so we should never reach this right?

Reply:
That's a good point, since we are still changing certificate pool frequently, I'll leave a panic here to see if we missed anything.